packages/integration-sdk-runtime/src/synchronization/index.ts

```diff
@@ -399,6 +414,84 @@ export async function uploadData<T extends UploadDataLookup, K extends keyof T>(
   );
 }
 
+/**
+ * Removes data from the rawData of the largest entity until the overall size
+ * of the data object is less than maxSize (defaulted to UPLOAD_SIZE_MAX).
+ *
+ * @param data
+ */
+export function shrinkRawData<T extends UploadDataLookup, K extends keyof T>(
+  data: T[K][],
+  logger,
+  maxSize = UPLOAD_SIZE_MAX,
+) {
+  const startTimeInMilliseconds = Date.now();
+  let totalSize = Buffer.byteLength(JSON.stringify(data));
+  const initialSize = totalSize;
+  let itemsRemoved = 0;
+
+  while (totalSize > maxSize) {
+    let largestEntityKey = '';
+    let largestEntitySize = 0;
+    let largestRawDataEntryKey = '';
+    let largestRawDataEntrySize = 0;
+    let largestItemKey = '';
+    let largestItemSize = 0;
+
+    // Find largest Entity
+    for (const entity in data) {
+      const length = JSON.stringify(data[entity]).length;
+      if (length > largestEntitySize) {
+        largestEntitySize = length;
+        largestEntityKey = entity;
+      }
+    }
+
+    // Find largest _rawData entry (typically 0, but check to be certain)
+    for (const rawEntry in data[largestEntityKey]['_rawData']) {
+      const length = JSON.stringify(
+        data[largestEntityKey]['_rawData'][rawEntry],
+      ).length;
+      if (length > largestRawDataEntrySize) {
+        largestRawDataEntrySize = length;
+        largestRawDataEntryKey = rawEntry;
+      }
+    }
+
+    // Find largest item within rawData
+    for (const item in data[largestEntityKey]['_rawData'][
+      largestRawDataEntryKey
+    ]['rawData']) {
+      const length = data[largestEntityKey]['_rawData'][largestRawDataEntryKey][
+        'rawData'
+      ][item]
+        ? JSON.stringify(
+            data[largestEntityKey]['_rawData'][largestRawDataEntryKey][
+              'rawData'
+            ][item],
+          ).length
+        : 0;
+      if (length > largestItemSize) {
+        largestItemKey = item;
+        largestItemSize = length;
+      }
+    }
+
+    data[largestEntityKey]['_rawData'][largestRawDataEntryKey]['rawData'][
+      largestItemKey
+    ] = 'TRUNCATED';
+    itemsRemoved += 1;
+    totalSize = Buffer.byteLength(JSON.stringify(data));
```

Comment:
Would the following be a constant equivalent:
```ts
    totalSize = totalSize - largestItemSize + 9
```
where 9 is the length of `TRUNCATED`

Reply:
Agreed with the overall idea for sure. Saving that many JSON.stringify calls for the entire data object should be some worthwhile savings.  Pushed an update to this in the latest commit.

Reply:
I think in the case of plain ASCII text the `.length` and the `.byteLength` will be the same.
```ts 
console.log("TRUNCATED".length === Buffer.byteLength("TRUNCATED")) // true
```
However UTF-8 characters like `🦄` will not have `.length` and `.byteLength` of the same size.
```ts
console.log("🦄".length === Buffer.byteLength("🦄")) // 2 === 4 -- false
console.log("ø".legnth === Buffer.byteLength("ø")) // 1 === 2 -- false
```
Currently, `largestItemSize` comes from the `.length` property of the string, but the `byteLength` of the request is what matters.

I think with the 16KB cushion built-in it may be safe 99% of the time, but occasionally, this could still produce a request larger than the 6MB limit if it contains a large number of UTF-8 characters. I'm not sure if `byteLength` is a constant time operation, but if it is then switching 
```ts
const length = data[largestEntityKey]['_rawData'][largestRawDataEntryKey][
        'rawData'
      ][item]
        ? JSON.stringify(
            data[largestEntityKey]['_rawData'][largestRawDataEntryKey][
              'rawData'
            ][item],
          ).length
        : 0;
```
to
```ts
const length = data[largestEntityKey]['_rawData'][largestRawDataEntryKey][
        'rawData'
      ][item]
        ? Buffer.byteLength(JSON.stringify(
            data[largestEntityKey]['_rawData'][largestRawDataEntryKey][
              'rawData'
            ][item],
          ))
        : 0;


```

Reply:
Thanks and good call out!  Agreed that it's probably safe, but I'm going to update it based on your recommendation all the same.  Overall still executing faster than before.